Userland/Services/WindowServer/ScreenBackend.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2018-2020, the SerenityOS developers.
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "ScreenBackend.h"
+#include "sys/ioctl_numbers.h"
+#include <AK/Try.h>
+#include <Kernel/API/FB.h>
+#include <LibCore/System.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+namespace WindowServer {
+
+HardwareScreenBackend::HardwareScreenBackend(String device)
+    : m_device(move(device))
+{
+}
+
+ErrorOr<void> HardwareScreenBackend::open()
```

Comment:
All screen backend implementations use a framebuffer, it's just what that framebuffer is connected to that's different. For hardware screens, the framebuffer is shared memory with the driver, for virtual screens, it's private malloc'ed memory.

Reply:
Well, right, but this is the implementation that opens `/dev/fbN` and uses ioctl to control the screen, rather than using an anonymous buffer. there's other ways to use a "hardware screen" than a framebuffer *device*. Probably moot until we actually implement such an alternative screen API though.